Hack: fix CreateDateOnlyFromDiscriminatorValue

Kiota has something funky going on with these structs, and we require a find/replace in post-processing to fix the package names used to reference them:

https://github.com/octokit/source-generator/blob/d1c29eea0c06262a42cc119b3ab0304f5e8e67ef/post-processors/go/main.go#L237-L252

When this step is removed, the error given is `versions/versions_request_builder.go:46:152: undefined: i878a80d2330e89d26896388a3f487eef27b0a0e6c010c493bf80be1452208f91.CreateDateOnlyFromDiscriminatorValue`. 

We should bring this up to the Kiota team and see if there's anything we can do to fix it.

--- a/.github/workflows/build-go.yml
+++ b/.github/workflows/build-go.yml
@@ -53,7 +53,7 @@ jobs:
           go build ./...
 
       - uses: gr2m/create-or-update-pull-request-action@cd-path
-        if: github.event_name != 'pull_request' # do not update the SDK on PR builds
+        # if: github.event_name != 'pull_request' # do not update the SDK on PR builds
         env:
           GITHUB_TOKEN: ${{ secrets.OCTOKITBOT_PAT }}
         with:

--- a/.github/workflows/build-go.yml
+++ b/.github/workflows/build-go.yml
@@ -53,7 +53,7 @@ jobs:
           go build ./...
 
       - uses: gr2m/create-or-update-pull-request-action@cd-path
-        # if: github.event_name != 'pull_request' # do not update the SDK on PR builds
+        if: github.event_name != 'pull_request' # do not update the SDK on PR builds
         env:
           GITHUB_TOKEN: ${{ secrets.OCTOKITBOT_PAT }}
         with:

--- a/.github/workflows/build-go.yml
+++ b/.github/workflows/build-go.yml
@@ -37,7 +37,7 @@ jobs:
         run: go run schemas/main.go --schema-next=false
 
       - name: Run generation
-        run: kiota generate -l go --ll trace -o $(pwd)/../go-sdk/github/octokit -n octokit -d schemas/downloaded.json
+        run: kiota generate -l go --ll trace -o $(pwd)/../go-sdk/github/octokit -n "github.com/octokit/go-sdk/github/octokit/" -d schemas/downloaded.json
 
       - name: Build post-processing
         run: go build -o post-processors/go/post-processor post-processors/go/main.go

--- a/.github/workflows/build-go.yml
+++ b/.github/workflows/build-go.yml
@@ -17,9 +17,9 @@ jobs:
       - name: Checkout repository
         uses: actions/checkout@v3
 
-      - uses: actions/setup-dotnet@v3.0.3
+      - uses: actions/setup-dotnet@v4
         with:
-          dotnet-version: '7.x.x'
+          dotnet-version: '8.x.x'
 
       - uses: actions/setup-go@v3.5.0
         with:

--- a/.github/workflows/build-go.yml
+++ b/.github/workflows/build-go.yml
@@ -26,7 +26,7 @@ jobs:
           go-version: '1.21.4'
 
       - name: Install kiota
-        run: dotnet tool install --global Microsoft.OpenApi.Kiota --version 1.8.2
+        run: dotnet tool install --global Microsoft.OpenApi.Kiota --prerelease
 
       - name: Clone the existing go-sdk
         run: |

--- a/README.md
+++ b/README.md
@@ -7,7 +7,7 @@ This repository generates code from [GitHub's OpenAPI specification](https://git
 - Install
 	- [.NET 7](https://dotnet.microsoft.com/en-us/download/dotnet/7.0)
 	- [Latest version of Go](https://go.dev/dl/)
-	- Kiota: `dotnet tool install --global Microsoft.OpenApi.Kiota --version 1.8.2`
+	- Kiota: `dotnet tool install --global Microsoft.OpenApi.Kiota --prerelease`
 - For the latest versions of tooling, the `.github/workflows/build-go.yml` file is the source of truth.
 
 ## Usage

--- a/post-processors/go/main.go
+++ b/post-processors/go/main.go
@@ -41,7 +41,6 @@ func run() error {
 			return fmt.Errorf("input file %v must not be empty", file.Name())
 		}
 
-		fileContents = fixImports(fileContents)
 		fileContents = fixKiotaNonDeterminism(fileContents, file.Name())
 
 		// TODO(kfcampbell): verify file permission is what we want
@@ -152,14 +151,6 @@ func walkFiles(path string, info fs.FileInfo, err error) error {
 	}
 	files[path] = info
 	return nil
-}
-
-// these fixes are working around bugs or limitations in Kiota and/or our schema
-func fixImports(inputFile string) string {
-	// find: go-sdk/
-	// replace: github.com/octokit/go-sdk/
-	inputFile = strings.ReplaceAll(inputFile, `"octokit/`, `"github.com/octokit/go-sdk/github/octokit/`)
-	return inputFile
 }
 
 // see https://github.com/microsoft/kiota/issues/3700

--- a/post-processors/go/main.go
+++ b/post-processors/go/main.go
@@ -42,7 +42,6 @@ func run() error {
 		}
 
 		fileContents = fixImports(fileContents)
-		fileContents = fixCreateDateOnlyFromDiscriminatorValue(fileContents, file.Name())
 		fileContents = fixKiotaNonDeterminism(fileContents, file.Name())
 
 		// TODO(kfcampbell): verify file permission is what we want
@@ -163,23 +162,7 @@ func fixImports(inputFile string) string {
 	return inputFile
 }
 
-func fixCreateDateOnlyFromDiscriminatorValue(inputFile string, filename string) string {
-	toReplace := `res, err := m.requestAdapter.SendCollection(ctx, requestInfo, i878a80d2330e89d26896388a3f487eef27b0a0e6c010c493bf80be1452208f91.CreateDateOnlyFromDiscriminatorValue, errorMapping)`
-	replaceWith := `res, err := m.requestAdapter.SendCollection(ctx, requestInfo, i158396662f32fe591e8faa247af18558546841dba91f24f5c824e11e34188830.CreateBasicErrorFromDiscriminatorValue, errorMapping)`
-	if strings.Contains(inputFile, toReplace) {
-		inputFile = strings.ReplaceAll(inputFile, toReplace, replaceWith)
-	}
-
-	toReplace = `res, err := m.BaseRequestBuilder.RequestAdapter.SendCollection(ctx, requestInfo, i878a80d2330e89d26896388a3f487eef27b0a0e6c010c493bf80be1452208f91.CreateDateOnlyFromDiscriminatorValue, errorMapping)`
-	replaceWith = `res, err := m.BaseRequestBuilder.RequestAdapter.SendCollection(ctx, requestInfo, i8bb20811a612dd15efa26f086111481a68f72cd9ac5da7a939a417131078d77e.CreateKeySimpleFromDiscriminatorValue, errorMapping)`
-
-	if strings.Contains(inputFile, toReplace) {
-		inputFile = strings.ReplaceAll(inputFile, toReplace, replaceWith)
-	}
-
-	return inputFile
-}
-
+// see https://github.com/microsoft/kiota/issues/3700
 func fixKiotaNonDeterminism(inputFile string, fileName string) string {
 	if !strings.Contains(fileName, "item_starred_repository.go") {
 		return inputFile

--- a/post-processors/go/main.go
+++ b/post-processors/go/main.go
@@ -43,7 +43,6 @@ func run() error {
 
 		fileContents = fixImports(fileContents)
 		fileContents = fixCreateDateOnlyFromDiscriminatorValue(fileContents, file.Name())
-		fileContents = dirtyHackForVersionsRequestBuilder(fileContents, file.Name())
 		fileContents = fixKiotaNonDeterminism(fileContents, file.Name())
 
 		// TODO(kfcampbell): verify file permission is what we want
@@ -161,36 +160,6 @@ func fixImports(inputFile string) string {
 	// find: go-sdk/
 	// replace: github.com/octokit/go-sdk/
 	inputFile = strings.ReplaceAll(inputFile, `"octokit/`, `"github.com/octokit/go-sdk/github/octokit/`)
-	return inputFile
-}
-
-func dirtyHackForVersionsRequestBuilder(inputFile string, fileName string) string {
-	if !strings.Contains(fileName, "versions_request_builder.go") {
-		return inputFile
-	}
-
-	toReplace := `
-	"github.com/microsoft/kiota-abstractions-go/serialization"`
-	replaceWith := ``
-
-	if strings.Contains(inputFile, toReplace) {
-		inputFile = strings.ReplaceAll(inputFile, toReplace, replaceWith)
-	}
-
-	toReplace = `for i, v := range res {
-        if v != nil {
-            val[i] = *(v.(*i878a80d2330e89d26896388a3f487eef27b0a0e6c010c493bf80be1452208f91.DateOnly))
-        }
-    }`
-	replaceWith = `// for i, v := range res {
-    //     if v != nil {
-    //         val[i] = *(v.(*i878a80d2330e89d26896388a3f487eef27b0a0e6c010c493bf80be1452208f91.DateOnly))
-    //     }`
-
-	if strings.Contains(inputFile, toReplace) {
-		inputFile = strings.ReplaceAll(inputFile, toReplace, replaceWith)
-	}
-
 	return inputFile
 }
 

--- a/scripts/install-tools.sh
+++ b/scripts/install-tools.sh
@@ -4,7 +4,7 @@
 KIOTA_TOOL_NAME="microsoft.openapi.kiota"
 # This version can only increase, never decrease. If you want to use a lower version, you need to uninstall the tool first.
 # dotnet tool uninstall Microsoft.OpenApi.Kiota --global
-KIOTA_TOOL_VERSION="1.8.2"
+KIOTA_TOOL_VERSION="v1.9.0-preview.202311300001"
 
 # Check if the tool is installed globally
 if dotnet tool list -g | grep -q $KIOTA_TOOL_NAME; then
@@ -12,10 +12,10 @@ if dotnet tool list -g | grep -q $KIOTA_TOOL_NAME; then
     echo "The correct version ($KIOTA_TOOL_VERSION) of $KIOTA_TOOL_NAME is installed...skipping install step"
   else
       echo "$KIOTA_TOOL_NAME is installed. Updating to sync with the required version"
-      dotnet tool update --global Microsoft.OpenApi.Kiota --version $KIOTA_TOOL_VERSION
+      dotnet tool update --global Microsoft.OpenApi.Kiota --prerelease
   fi
 else
   echo "$KIOTA_TOOL_NAME is not installed...installing"
-  dotnet tool install --global Microsoft.OpenApi.Kiota --version $KIOTA_TOOL_VERSION
+  dotnet tool install --global Microsoft.OpenApi.Kiota --prerelease
 fi
 

